File: No.22_GenerateParentheses/generateParenthesis.java
class Solution {
	public List<String> generateParenthesis(int n){
        List<String> ans = new ArrayList();
        recursive(ans, "", 0, 0, n);
        return ans;
        
    }
	private void recursive(List<String> ans, String s, int left, int right, int n) {
		if (s.length()==2*n)
		{
			ans.add(s);
			return;
		}
		if (left<n)
			recursive(ans, s+"(", left+1, right, n);
		if (right<left)
			recursive(ans, s+")", left, right+1, n);
	}
}

File: No.22_GenerateParentheses/generateParenthesis.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        ans = []
        def recursive(s='', left = 0, right = 0):
            if len(s) == n*2:
                ans.append(s)
                return
            if left < n:
                recursive(s+'(', left+1, right)
            if right < left:
                recursive(s+')', left, right+1)
                
        recursive()
        return ans